src/main/java/org/apache/maven/plugins/install/InstallMojo.java

```diff
@@ -91,53 +86,68 @@
     public void execute()
         throws MojoExecutionException, MojoFailureException
     {
-        boolean addedInstallRequest = false;
         if ( skip )
         {
+            getPluginContext().put( INSTALL_PROCESSED_MARKER, Boolean.FALSE );
             getLog().info( "Skipping artifact installation" );
         }
         else
         {
-            // CHECKSTYLE_OFF: LineLength
-            ProjectInstallerRequest projectInstallerRequest =
-                new ProjectInstallerRequest().setProject( project );
-            // CHECKSTYLE_ON: LineLength
-
             if ( !installAtEnd )
             {
-                installProject( session.getProjectBuildingRequest(), projectInstallerRequest );
+                installProject( project );
             }
             else
             {
-                INSTALLREQUESTS.add( projectInstallerRequest );
-                addedInstallRequest = true;
+                getPluginContext().put( INSTALL_PROCESSED_MARKER, Boolean.TRUE );
+                getLog().info( "Installing " + getProjectReferenceId( project ) + " at end" );
             }
         }
 
-        boolean projectsReady = READYPROJECTSCOUNTER.incrementAndGet() == reactorProjects.size();
-        if ( projectsReady )
+        if ( allProjectsMarked() )
         {
-            synchronized ( INSTALLREQUESTS )
+            for ( MavenProject reactorProject : reactorProjects )
             {
-                while ( !INSTALLREQUESTS.isEmpty() )
+                Map<String, Object> pluginContext = session.getPluginContext( pluginDescriptor, reactorProject );
+                Boolean install = (Boolean) pluginContext.get( INSTALL_PROCESSED_MARKER );
```

Comment:
Can this be `null` at any time?

Reply:
no, as it is here, the `allProjectsMarked()` returned true, hence, all the modules has been marked (have value set)